private/buf/cmd/buf/command/policy/policyprune/policyprune.go

```diff
@@ -0,0 +1,114 @@
+	"github.com/bufbuild/buf/private/buf/bufcli"
+	"github.com/bufbuild/buf/private/buf/bufworkspace"
+	"github.com/bufbuild/buf/private/bufpkg/bufparse"
+	"github.com/bufbuild/buf/private/bufpkg/bufpolicy"
+)
+
+// NewCommand returns a new Command.
+func NewCommand(
+	name string,
+	builder appext.SubCommandBuilder,
+) *appcmd.Command {
+	return &appcmd.Command{
+		Use:   name + " <directory>",
+		Short: "Prune unused policies from buf.lock",
+		Long: `Policies that are no longer configured in buf.yaml are removed from the buf.lock file.
+
+The first argument is the directory of your buf.yaml configuration file.
+Defaults to "." if no argument is specified.`,
+		Args: appcmd.MaximumNArgs(1),
+		Run: builder.NewRunFunc(
+			func(ctx context.Context, container appext.Container) error {
+				return run(ctx, container)
+			},
+		),
+	}
+}
+
+func run(
+	ctx context.Context,
+	container appext.Container,
+) error {
+	dirPath := "."
+	if container.NumArgs() > 0 {
+		dirPath = container.Arg(0)
+	}
+	controller, err := bufcli.NewController(container)
+	if err != nil {
+		return err
+	}
+	workspaceDepManager, err := controller.GetWorkspaceDepManager(ctx, dirPath)
+	if err != nil {
+		return err
+	}
+	configuredRemotePolicyRefs, err := workspaceDepManager.ConfiguredRemotePolicyRefs(ctx)
+	if err != nil {
+		return err
+	}
+	return prune(
+		ctx,
+		xslices.Map(
+			configuredRemotePolicyRefs,
+			func(policyRef bufparse.Ref) string {
+				return policyRef.FullName().String()
+			},
+		),
+		workspaceDepManager,
+	)
+}
+
+func prune(
+	ctx context.Context,
+	bufYAMLBasedRemotePolicyNames []string,
+	workspaceDepManager bufworkspace.WorkspaceDepManager,
+) error {
+	bufYAMLRemotePolicyNames := xslices.ToStructMap(bufYAMLBasedRemotePolicyNames)
+	existingRemotePolicyKeys, err := workspaceDepManager.ExistingBufLockFileRemotePolicyKeys(ctx)
+	if err != nil {
+		return err
+	}
+	var prunedBufLockPolicyKeys []bufpolicy.PolicyKey
+	for _, existingRemotePolicyKey := range existingRemotePolicyKeys {
+		// Check if an existing plugin key from the buf.lock is configured in the buf.yaml.
+		if _, ok := bufYAMLRemotePolicyNames[existingRemotePolicyKey.FullName().String()]; ok {
+			// If yes, then we keep it for the updated buf.lock.
+			prunedBufLockPolicyKeys = append(prunedBufLockPolicyKeys, existingRemotePolicyKey)
+		}
+	}
+	// We keep the existing dep module keys as-is.
+	existingDepModuleKeys, err := workspaceDepManager.ExistingBufLockFileDepModuleKeys(ctx)
+	if err != nil {
+		return err
+	}
+	existingRemotePluginKeys, err := workspaceDepManager.ExistingBufLockFileRemotePluginKeys(ctx)
+	if err != nil {
+		return err
+	}
+	existingPolicyNameToRemotePolicyKeys, err := workspaceDepManager.ExistingBufLockFilePolicyNameToRemotePluginKeys(ctx)
+	if err != nil {
+		return err
+	}
+	return workspaceDepManager.UpdateBufLockFile(ctx, existingDepModuleKeys, existingRemotePluginKeys, prunedBufLockPolicyKeys, existingPolicyNameToRemotePolicyKeys)
```

Comment:
Shouldn't we also be pruning the `policyNameToRemotePolicyKeys`?

Reply:
Thanks, fixed.